src/compute/src/compute_state.rs

```diff
@@ -129,6 +131,25 @@ pub struct ComputeState {
     ///
     /// Copies of this sender are passed to the hydration logging operators.
     pub hydration_tx: mpsc::Sender<HydrationEvent>,
+
+    /// Dataflows in the process of hydrating.
+    ///
+    /// Each entry is a list of IDs of collections exported by a dataflow. The dataflow is fully
+    /// hydrated when all exported collections are hydrated.
+    hydrating_dataflows: BTreeSet<Vec<GlobalId>>,
+    /// Queue of dataflows awaiting hydration.
+    ///
+    /// Each entry is a list of collection IDs exported by a dataflow, and a token that can be
+    /// dropped to unsuspend the dataflow.
+    hydration_queue: VecDeque<(Vec<GlobalId>, Box<dyn Any>)>,
+    /// A watch receiver to be used during rendering to control suspension of data-producing
+    /// operators while they await hydration.
+    ///
+    /// If `Some`, data-producing operators should remain suspended as long as the corresponding
+    /// watch sender has not hung up.
+    ///
+    /// Note that currently only `persist_source` supports suspension, arrangement imports do not.
+    hydration_watch: Option<watch::Receiver<()>>,
```

Comment:
Yeah, I'd hoped to avoid some boilerplate but you are right that this obscures the control flow. Definitely prefer to add a bit more code than trying to explain less code with more comments :)